crates/uv/src/commands/version.rs

```diff
@@ -1,20 +1,197 @@
-use anyhow::Result;
+use std::fmt::Write;
+use std::str::FromStr;
+use std::{cmp::Ordering, path::Path};
 
-use uv_cli::VersionFormat;
+use anyhow::{anyhow, Result};
+use owo_colors::OwoColorize;
 
-/// Display version information
-pub(crate) fn version(output_format: VersionFormat, buffer: &mut dyn std::io::Write) -> Result<()> {
-    let version_info = uv_cli::version::version();
+use uv_cli::version::VersionInfo;
+use uv_cli::{VersionBump, VersionFormat};
+use uv_fs::Simplified;
+use uv_pep440::Version;
+use uv_workspace::pyproject_mut::Error;
+use uv_workspace::{
+    pyproject_mut::{DependencyTarget, PyProjectTomlMut},
+    DiscoveryOptions, Workspace, WorkspaceCache,
+};
 
+use crate::{commands::ExitStatus, printer::Printer};
+
+/// Display version information for uv itself
+pub(crate) fn self_version(
+    short: bool,
+    output_format: VersionFormat,
+    printer: Printer,
+) -> Result<ExitStatus> {
+    let version_info = uv_cli::version::uv_self_version();
+    print_version(version_info, None, short, output_format, printer)?;
+
+    Ok(ExitStatus::Success)
+}
+
+/// Read or update project version (`uv metadata version`)
+pub(crate) async fn project_version(
+    project_dir: &Path,
+    value: Option<String>,
+    bump: Option<VersionBump>,
+    dry_run: bool,
+    short: bool,
+    output_format: VersionFormat,
+    strict: bool,
+    cache: &WorkspaceCache,
+    printer: Printer,
+) -> Result<ExitStatus> {
+    // Read the metadata
+    let workspace =
+        match Workspace::discover(project_dir, &DiscoveryOptions::default(), cache).await {
+            Ok(workspace) => workspace,
+            Err(err) => {
+                // If strict, hard bail on missing workspace
+                if strict {
+                    return Err(err)?;
+                }
+                // Otherwise, warn and provide fallback
+                writeln!(
+                    printer.stderr(),
+                    r"warning: failed to read project: {err}
+  running `uv self version` for compatibility with old `uv version` command.
+  this fallback will be removed soon, pass `--preview` to make this an error.
+"
+                )?;
+                return self_version(short, output_format, printer);
+            }
+        };
+
+    let mut pyproject = PyProjectTomlMut::from_toml(
+        &workspace.pyproject_toml().raw,
+        DependencyTarget::PyProjectToml,
+    )?;
+    let pyproject_path = workspace.install_path().join("pyproject.toml");
+    let name = workspace
+        .pyproject_toml()
+        .project
+        .as_ref()
+        .map(|project| &project.name);
+    let old_version = pyproject.version().map_err(|err| match err {
+        Error::MalformedWorkspace => {
+            if pyproject.has_dynamic_version() {
+                anyhow!(
+                    "We cannot get or set dynamic project versions in: {}",
```

Comment:
nit: backticks around the path

Reply:
We don't use backticks following a `:` usually